src/test/java/io/netty/buffer/api/examples/bytetomessagedecoder/ByteToMessageDecoderTest.java

```diff
@@ -0,0 +1,561 @@
+    }
+
+    @Test
+    public void testDecodeLastNonEmptyBuffer() {
+        EmbeddedChannel channel = new EmbeddedChannel(new ByteToMessageDecoder() {
+            private boolean decodeLast;
+
+            @Override
+            protected void decode(ChannelHandlerContext ctx, Buffer in) {
+                int readable = in.readableBytes();
+                assertTrue(readable > 0);
+                if (!decodeLast && readable == 1) {
+                    return;
+                }
+                int read = decodeLast ? readable : readable - 1;
+                Buffer slice = in.slice(in.readerOffset(), read);
+                in.readerOffset(in.readerOffset() + read);
+                ctx.fireChannelRead(slice);
+            }
+
+            @Override
+            protected void decodeLast(ChannelHandlerContext ctx, Buffer in) throws Exception {
+                assertFalse(decodeLast);
+                decodeLast = true;
+                super.decodeLast(ctx, in);
+            }
+        });
+        byte[] bytes = new byte[1024];
+        ThreadLocalRandom.current().nextBytes(bytes);
+        try (Buffer buf = heap().allocate(bytes.length, BIG_ENDIAN);
+             Buffer part1 = buf.slice(0, bytes.length - 1);
+             Buffer part2 = buf.slice(bytes.length - 1, 1)) {
+            for (byte b : bytes) {
+                buf.writeByte(b);
+            }
+            assertTrue(channel.writeInbound(buf.slice()));
+            try (Buffer actual = channel.readInbound()) {
+                assertEquals(part1, actual);
+            }
+            assertNull(channel.readInbound());
+            assertTrue(channel.finish());
+            try (Buffer actual = channel.readInbound()) {
+                assertEquals(part2, actual);
+            }
+            assertNull(channel.readInbound());
+        }
+    }
+
+    @Test
+    public void testReadOnlyBuffer() {
+        EmbeddedChannel channel = new EmbeddedChannel(new ByteToMessageDecoder() {
+            @Override
+            protected void decode(ChannelHandlerContext ctx, Buffer in) { }
+        });
+        assertFalse(channel.writeInbound(heap().allocate(8).writeByte((byte) 1).readOnly(true)));
+        assertFalse(channel.writeInbound(heap().allocate(1).writeByte((byte) 2)));
+        assertFalse(channel.finish());
+    }
+
+    static class WriteFailingByteBuf extends BufferAdaptor {
+        private final Error error = new Error();
+        private int untilFailure;
+
+        WriteFailingByteBuf(int untilFailure, int capacity) {
+            this(untilFailure, heap().allocate(capacity, BIG_ENDIAN));
+            this.untilFailure = untilFailure;
+        }
+
+        private WriteFailingByteBuf(int untilFailure, Buffer buffer) {
+            super(buffer);
+            this.untilFailure = untilFailure;
+        }
+
+        @Override
+        public Buffer order(ByteOrder order) {
+            if (order == LITTLE_ENDIAN && --untilFailure <= 0) {
+                throw error;
```

Comment:
This seems like a weird need for `BufferAdaptor`. I thought we were going with the idea "people generally wouldn't implement their own Buffer." If that's the case, then injecting an exception into this simple setter seems unnecessary, as it simply wouldn't happen. But if we introduce something like BufferAdaptor, that _encourages_ people to create their own Buffer.

Reply:
The test was doing something like that, and I wanted to port it over as faithfully as possible. But you're right. We shouldn't have something like that, and the test needs to be adapted, which I can do now that I know it passes as written.